### PR TITLE
CORE-8881 Maintain the allowed set of TLS certificate subjects

### PR DESCRIPTION
**Pull request by williamvigorr3**
Read the certificate subjects published by the Link Manager inside the gateway. Merge them from the 3 sources into a single source from which we can do a lookup by certificate subject.

### CODE REVIEW

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStore.kt [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import net.corda.lifecycle.domino.logic.LifecycleWithDominoTile
+import net.corda.lifecycle.domino.logic.util.SubscriptionDominoTile
+import net.corda.messaging.api.processor.CompactedProcessor
+import net.corda.messaging.api.records.Record
+import net.corda.messaging.api.subscription.config.SubscriptionConfig
+import net.corda.messaging.api.subscription.factory.SubscriptionFactory
+import net.corda.schema.Schemas
+import java.util.concurrent.ConcurrentHashMap
+
+class DynamicCertificateSubjectStore(
```
Comment (yift-r3) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStore.kt]: Can we make it internal?

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStore.kt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    override val dominoTile = SubscriptionDominoTile(
+        lifecycleCoordinatorFactory,
+        subscription,
+        subscriptionConfig,
+        emptyList(),
+        emptyList(),
+    )
+
+    fun subjectAllowed(subject: CertificateSubject): Boolean {
+        return certificateSubjects[subject] != null
```
Comment (yift-r3) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStore.kt]: containsKey might be faster

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStore.kt [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    private fun addSource(subject: CertificateSubject, source: String) {
+        certificateSubjects.compute(subject) { _, sources ->
+            sources?.apply { add(source) } ?: mutableSetOf(source)
```
Comment (yift-r3) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStore.kt:47]: Can we make it a concurrent set (ConcurrentHashMap.newKeySet)?

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStore.kt [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        }
+    }
+
+    private fun removeSource(subject: CertificateSubject, source: String) {
+        certificateSubjects.compute(subject) { _, sources ->
```
Comment (yift-r3) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStore.kt]: Maybe `computeIfPresent`?

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStore.kt (hunk 3 above)
Comment (williamvigorr3) [reply components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStore.kt:47]: This should be thread safe as we are doing a compute on the outer concurrent hash map.

**Review by yift-r3 — DISMISSED** [review]

**Review by yift-r3 — APPROVED** [review]

**Pull request merged**